src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -0,0 +1,3129 @@
+            generatedName += ComputeStringHash(generatedName).ToString("X");
+
+            // Generate the regex type
+            GenerateRegex(writer, regexClass.Method, generatedName);
+
+            while (writer.Indent != 0)
+            {
+                writer.Indent--;
+                writer.WriteLine("}");
+            }
+
+            // FNV-1a hash function.  The actual algorithm used doesn't matter; just something simple
+            // to create a pseudo-random value based on input text.
+            static uint ComputeStringHash(string s)
+            {
+                uint hashCode = 2166136261;
+                foreach (char c in s)
+                {
+                    hashCode = (c ^ hashCode) * 16777619;
+                }
+                return hashCode;
+            }
+        }
+
+        /// <summary>Gets whether a given regular expression method is supported by the code generator.</summary>
+        private static bool SupportsCustomCodeGeneration(RegexMethod rm)
+        {
+            const RegexOptions SupportedOptions =
+                RegexOptions.IgnoreCase |
+                RegexOptions.Multiline |
+                RegexOptions.ExplicitCapture |
+                RegexOptions.Compiled |
+                RegexOptions.Singleline |
+                RegexOptions.IgnorePatternWhitespace |
+                RegexOptions.RightToLeft |
+                RegexOptions.ECMAScript |
+                RegexOptions.CultureInvariant;
+
+            // If we see an option we're not aware of (but that was allowed through), don't emit custom regex code.
+            return (rm.Options & ~(int)SupportedOptions) == 0;
+        }
+
+        /// <summary>Generates the code for a regular expression method.</summary>
+        private static void GenerateRegex(IndentedTextWriter writer, RegexMethod rm, string id)
+        {
+            string patternExpression = Literal(rm.Pattern);
+            string optionsExpression = $"(RegexOptions)({rm.Options})";
+            string timeoutExpression = rm.MatchTimeout == Timeout.Infinite ?
+                "Timeout.InfiniteTimeSpan" :
+                $"TimeSpan.FromMilliseconds({rm.MatchTimeout.Value.ToString(CultureInfo.InvariantCulture)})";
+
+            writer.WriteLine(s_generatedCodeAttribute);
+            writer.WriteLine($"{rm.Modifiers} Regex {rm.MethodName}() => {id}.Instance;");
+            writer.WriteLine();
+            writer.WriteLine(s_generatedCodeAttribute);
+            writer.WriteLine($"{(writer.Indent != 0 ? "private" : "internal")} sealed class {id} : Regex");
+            writer.WriteLine("{");
+            writer.Write($"    public static Regex Instance {{ get; }} = ");
+
+            // If we can't support custom generation for this regex, spit out a Regex constructor call.
+            if (!SupportsCustomCodeGeneration(rm))
+            {
+                writer.WriteLine($"new Regex({patternExpression}, {optionsExpression}, {timeoutExpression});");
```

Comment:
should we have a diagnostic for this? they explicitly asked for the generator using the attribute, it's not that we are opportunistically discovering them.

Reply:
We can decide whether to fail, having a warning diagnostic but still proceed, or just fall back to this.  This won't actually be used right now, as we support all options that make it to this point; I added this so that we don't just fail when NonBacktracking gets added in, just as we don't want RegexOptions.Compiled | RegexOptions.NonBacktracking to start failing.